phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java

```diff
@@ -0,0 +1,422 @@
+            }
+        }
+    }
+
+    private void updateResultsWithRowOutsideTTLWindow(List<Cell> result,
+            CompactionRowVersion rowVersion) {
+        if (rowVersion.deleteFamilyMarker == null) {
+            if (firstStore) {
+                if (rowVersion.version < minVersion) {
+                    // Live rows whose version is less than min version are retained outside
+                    // the TTL window
+                    addCells(rowVersion, result);
+                }
+            } else if (rowVersion.version < maxVersion) {
+                // check if this row needs to be retained
+            } else if (keepDeletedCells == KeepDeletedCells.TTL) {
+                if (rowVersion.deleteFamilyMarker.getTimestamp() >= ttlWindowStart) {
+                    addCells(rowVersion, result);
+                }
+            }
+        }
+    }
+    private void prepareResults(List<Cell> result, List<CompactionRowVersion> rows,
+            List<Cell> deleteMarkers) {
+        for (CompactionRowVersion row : rows) {
+            if (row.ts >= maxLookbackWindowStart) {
+                // All rows within the max lookback window are retained
+                result.addAll(row.cells);
+            }
+            else if (row.ts >= ttlWindowStart) {
+                updateResultsWithRowOutsideMaxLookbackButInsideTTLWindow(result, row);
+            } else {
+                updateResultsWithRowOutsideTTLWindow(result, row);
+            }
+        }
+        for (Cell cell : deleteMarkers) {
+            if (cell.getTimestamp() >= maxLookbackWindowStart) {
+                result.add(cell);
+            }
+            else if (cell.getTimestamp() >= ttlWindowStart) {
+                if (keepDeletedCells != KeepDeletedCells.FALSE) {
+                    result.add(cell);
+                }
+            }
+        }
+    }
+    /**
+     * Filter {@link Cell}s from the underlying store scanner
+     *
+     * @param result
+     *            next batch of {@link Cell}s
+     * @throws IOException 
+     */
+    private void filter(List<Cell> result, boolean storeLevel) throws IOException {
+        if (result.isEmpty()) {
+            return;
+        }
+        List<List<Cell>> columns = new ArrayList<>();
+        List<Cell> deleteMarkers = new ArrayList<>();
+        formColumns(result, columns, deleteMarkers);
+        List<CompactionRowVersion> rows = new ArrayList<>();
+        if (formRowVersions(columns, rows, storeLevel)) {
+            result.clear();
+            prepareResults(result, rows, deleteMarkers);
+        } else {
+            filterRegionLevel(result);
+        }
+    }
+
+    private void filterRegionLevel(List<Cell> result) throws IOException {
+        Scan scan = new Scan();
+        scan.setRaw(true);;
+        scan.readAllVersions();
+        scan.setTimeRange(0, compactionTime);
+        byte[] rowKey = result.get(0).getRowArray();
+        scan.withStartRow(rowKey, true);
```

Comment:
calling scan.withStartRow twice? Is the second one supposed to be scan.withStopRow(rowKey, true)?

Reply:
Good catch! Yes, the second one is for the stop row.